bottomless2/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+/// receiver is dropped, and the queue is empty, exit.
+pub struct BottomlessLoop<S: Storage> {
+    receiver: mpsc::Receiver<StoreSegmentRequest<S::Config>>,
+    scheduler: Scheduler<S>,
+    max_in_flight: usize,
+    in_flight_futs: JoinSet<JobResult<S>>,
+    force_shutdown: oneshot::Receiver<()>,
+}
+
+impl<S: Storage + 'static> BottomlessLoop<S> {
+    /// Schedules durability jobs. This loop is not allowed to fail, or loose jobs.
+    /// A job is prepared by calling `Scheduler::prepare(..)`. The job is spawned, and it returns a
+    /// `JobResult`, which is then returned to the scheduler by calling `Scheduuler::report(..)`.
+    /// When a request is received, it is immediately scheduled by calling `Scheduler::register`
+    /// with it.
+    ///
+    /// The loop is only allowed to shutdown if the receiver is closed, and the scheduler is empty,
+    /// or if `force_shutdown` is called, in which case everything is dropped in place.
+    #[tracing::instrument(skip(self))]
+    async fn run(mut self) {
+        let mut shutting_down = false;
+        // run the loop until shutdown.
+        loop {
+            if shutting_down && self.scheduler.is_empty() {
+                break;
+            }
+
+            // schedule as much work as possible
+            while self.scheduler.has_work() && self.in_flight_futs.len() < self.max_in_flight {
+                let job = self
+                    .scheduler
+                    .schedule()
+                    .expect("scheduler has work, but didn't return a job");
+                self.in_flight_futs.spawn(job.perform());
+            }
+
+            tokio::select! {
+                msg = self.receiver.recv(), if !shutting_down => {
+                    match msg {
+                        Some(req) => {
+                            self.scheduler.register(req);
+                        }
+                        None => {
+                            shutting_down = true;
+                        }
+                    }
+                }
+                Some(join_result) = self.in_flight_futs.join_next(), if !self.in_flight_futs.is_empty() => {
+                    match join_result {
+                        Ok(job_result) => {
+                            // if shutting down, log progess:
+                            if shutting_down {
+                                tracing::info!("processed job, {} jobs remaining", self.in_flight_futs.len());
+                            }
+                            self.scheduler.report(job_result);
+                        }
+                        Err(e) => {
+                            // job panicked. report and exit process. The program is crippled, from
+                            // now on, so we just exit, and hope to restart on a fresh state.
+                            tracing::error!("fatal error: bottomless job panicked: {e}");
+                            std::process::exit(1);
```

Comment:
should we make this choice at a higher layer? Seems a bit extreme here to abort the whole process at a layer this deep.